c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -694,11 +694,244 @@ bool parse_body(JSContext *cx, HandleObject self, UniqueChars buf, size_t len) {
+    return false;
+  }
+  MOZ_ASSERT(foundContents);
+#endif
+  RootedValue contents_val(cx);
+  if (!JS_GetElement(cx, catch_handler, 1, &contents_val)) {
+    return false;
+  }
+  MOZ_ASSERT(contents_val.isObject());
+  RootedObject contents(cx, &contents_val.toObject());
+  if (!contents) {
+    return false;
+  }
+#ifdef DEBUG
+  bool contentsIsArray;
+  if (!JS::IsArrayObject(cx, contents, &contentsIsArray)) {
+    return false;
+  }
+  MOZ_ASSERT(contentsIsArray);
+#endif
+
+  auto reader_val = js::GetFunctionNativeReserved(catch_handler, 1);
+  MOZ_ASSERT(reader_val.isObject());
+  RootedObject reader(cx, &reader_val.toObject());
+
+  // TODO: Confirm this -- I think it is not true...
+  // We're guaranteed to work with a native ReadableStreamDefaultReader here,
+  // which in turn is guaranteed to vend {done: bool, value: any} objects to
+  // read promise then callbacks.
+  MOZ_ASSERT(args[0].isObject());
+  RootedObject chunk_obj(cx, &args[0].toObject());
+  RootedValue done_val(cx);
+  RootedValue value(cx);
+#ifdef DEBUG
+  bool hasValue;
+  if (!JS_HasProperty(cx, chunk_obj, "value", &hasValue)) {
+    return false;
+  }
+  MOZ_ASSERT(hasValue);
+#endif
+  if (!JS_GetProperty(cx, chunk_obj, "value", &value)) {
+    return false;
+  }
+#ifdef DEBUG
+  bool hasDone;
+  if (!JS_HasProperty(cx, chunk_obj, "done", &hasDone)) {
+    return false;
+  }
+  MOZ_ASSERT(hasDone);
+#endif
+  if (!JS_GetProperty(cx, chunk_obj, "done", &done_val)) {
+    return false;
+  }
+  MOZ_ASSERT(done_val.isBoolean());
+  if (done_val.toBoolean()) {
+    // We finished reading the stream
+    // Now we need to iterate/reduce `contents` JS Array into UniqueChars
+    uint32_t contentsLength;
+    if (!JS::GetArrayLength(cx, contents, &contentsLength)) {
+      return false;
+    }
+    std::vector<uint8_t> allBytes;
+    for (uint32_t index = 0; index < contentsLength; index++) {
+      RootedValue val(cx);
+      if (!JS_GetElement(cx, contents, index, &val)) {
+        return false;
+      }
+      {
+        JS::AutoCheckCannotGC nogc;
+        MOZ_ASSERT(val.isObject());
+        JSObject *array = &val.toObject();
+        MOZ_ASSERT(JS_IsUint8Array(array));
+        bool is_shared;
+        uint8_t *bytes = JS_GetUint8ArrayData(array, &is_shared, nogc);
+        size_t length = JS_GetTypedArrayByteLength(array);
+        allBytes.insert(allBytes.end(), bytes, bytes + length);
+      }
+    }
+#ifdef DEBUG
+    bool foundBodyParser;
+    if (!JS_HasElement(cx, catch_handler, 2, &foundBodyParser)) {
+      return false;
+    }
+    MOZ_ASSERT(foundBodyParser);
+#endif
+    // Now we can call parse_body on the result
+    RootedValue body_parser(cx);
+    if (!JS_GetElement(cx, catch_handler, 2, &body_parser)) {
+      return false;
+    }
+    auto parse_body = (ParseBodyCB *)body_parser.toPrivate();
+    size_t bytes_read = allBytes.size();
+    UniqueChars buf(reinterpret_cast<char *>(allBytes.data()));
```

Comment:
I'm not sure if this is safe: `UniqueChars` is a unique pointer holding a pointer to `char []` data, and will free the data when its destructor runs. As the vector still holds the bytes, this will likely be a double-free.

Reply:
I believe this is safe as we transfer ownership from the `UniqueChars buf` via the `std::move(buf)` call - is this not the case?

https://searchfox.org/mozilla-central/source/mfbt/UniquePtr.h#63-64

Reply:
The problem is that `allBytes` still owns the memory when `buf` is constructed. When `buf` is destroyed in `parse_body` it will call `free` on the bytes associated with `allBytes.data()`, and when `allBytes` is destroyed when this function returns it will also try to free the bytes.

Reply:
After reading through the `parse_body` implementation once more I can see the double free 👍 
I've refactored this to no longer use std::vector and instead use a char* which removes the double free issue